boards/pygamer/src/pins.rs

```diff
@@ -581,30 +582,54 @@ pub struct Display {
     pub tft_backlight: TftBacklightReset,
 }
 
+/// This empty error occurs if there is an issue setting up the on-board
+/// display.
+#[derive(Debug)]
+pub struct DisplayError;
+impl From<()> for DisplayError {
+    fn from(_value: ()) -> Self {
+        DisplayError
+    }
+}
+
```

Comment:
I feel like this is not descriptive enough. It does tell the user that an error occured, but it doesn't help determining what failed or why. Plus, it seems like all errors that may occur are of the same type, except for two of them that come from completely different sources:

* Setting up the clocks:
   Typically, the BSPs unwrap on an error when trying to setup a `sercom_*` or `tc*` clock. I don't think there's really a reasonable way to recover from this anyway, so `unwrap`ping is completely fine. Or better yet, using `expect`.

* Initializing the `ExclusiveDevice` struct:
   I seems like the only way `ExclusiveDevice::new_no_delay` can fail is if there's an error setting the CS pin to HIGH on initialization. For `atsamd-hal`, setting the state of an output pin is infaillible; this error will never happen. Again, I think `unwrap`ping or `expect` ing is perfectly ok here.

That means that the user will know for sure that the error comes from the display without having to add variants to your error enum, and not some other, miscellanous thing occured, like the clocks having been setup wrong.

It's just a personal preference though, and there's nothing inherently wrong with this approach, so feel fre to apply my suggestion or not.

Reply:
So, previously, `pygamer::pins::Display::init` returned a `Result<_, ()>`. This now triggers [this clippy lint](https://rust-lang.github.io/rust-clippy/master/index.html#result_unit_err), so, in the interest of not introducing new clippy lints (technically I suppose it wasn't my changes that triggered this), I added the empty `DisplayError` type.

So we have three options for the return type of `pygamer::pins::Display::init`:
1. Return the `Result<_, DisplayError>` as it currently does.
2. Keep it the way it was, returning `Result<_, ()>` and suppress the lint.
3. Simply return the pair `(DisplayDriver, Pwm2<PA01>)` instead of a `Result` and unwrap/panic the two errors above.

I have no particular preference here, but note that options 1 and 3 are breaking API changes, not that that matters too much since `pygamer` is not yet at version 1.0.0. I would comment though that, since this is library code, it might be best to let the end user decide whether failing to setup the `sercom` clock (and therefore the display) is recoverable or not. They could also just want to handle that problem in a way different from a panic.

What are everyone's thoughts here?

Reply:
Again, I don't really have a dog in the fight, and I don't want to start bikeshedding here. I'm just saying that as a user, I would be surprised to see a `DisplayError` which is actually a clock setup error. I would then suggest changing the `DisplayError` definition to something similar to this:

```rust
pub enum DisplayError {
    SercomClockMisconfigured,
    DriverError,
}
```

which is more in line with the principle of least surprise. I will say though that I'm surprised that all the functions in `st7735-lcd` return `Result<(), ()>`. That API certainly isn't the best at helping their users figure out what exactly went wrong.

Reply:
😅